Makefile

```diff
@@ -1,29 +1,159 @@
-TARGET_BPF := test/test.bpf.o
-VMLINUX_H = test/vmlinux.h
+BASEDIR = $(abspath ./)
 
-GO_SRC := $(shell find . -type f -name '*.go')
-BPF_SRC := $(shell find . -type f -name '*.bpf.c')
-PWD := $(shell pwd)
+OUTPUT = ./output
+SELFTEST = ./selftest
 
-LIBBPF_HEADERS := /usr/include/bpf
-LIBBPF := "-lbpf"
+CC = gcc
+CLANG = clang
 
-.PHONY: all
-all: test
+BTFFILE = /sys/kernel/btf/vmlinux
+BPFTOOL = $(shell which bpftool || /bin/false)
+VMLINUX_H = $(abspath $(OUTPUT)/vmlinux.h)
 
-$(VMLINUX_H):
-	bpftool btf dump file /sys/kernel/btf/vmlinux format c > test/vmlinux.h
+# libbpf
 
-go_env := CC=gcc CGO_CFLAGS="-I $(LIBBPF_HEADERS)" CGO_LDFLAGS="$(LIBBPF)"
-.PHONY: test
-test: $(TARGET_BPF) $(GO_SRC)
-	$(go_env) go test -ldflags '-extldflags "-static"' .
+LIBBPF_SRC = $(abspath ./libbpf/src)
+LIBBPF_OBJ = $(abspath ./$(OUTPUT)/libbpf.a)
+LIBBPF_OBJDIR = $(abspath ./$(OUTPUT)/libbpf)
+LIBBPF_DESTDIR = $(abspath ./$(OUTPUT))
 
-$(TARGET_BPF): $(BPF_SRC) $(VMLINUX_H)
-	clang \
-		-g -O2 -c -target bpf \
-		-o $@ $<
+CFLAGS = -g -O2 -Wall -fpie
+LDFLAGS =
 
-.PHONY: clean
-clean: 
-	rm $(TARGET_BPF) $(VMLINUX_H)
+# golang
+
+CGO_CFLAGS_STATIC = "-I$(abspath $(OUTPUT))"
+CGO_LDFLAGS_STATIC = "-lelf -lz $(LIBBPF_OBJ)"
+
+CGO_CFGLAGS_DYN = "-I. -I/usr/include/"
+CGO_LDFLAGS_DYN = "-lelf -lz -lbpf"
+
+# default == shared lib from OS package
+
+all: libbpfgo-dynamic
+install: libbpfgo-dynamic-install
+test: libbpfgo-dynamic-test
+
+# libbpfgo test object
+
+libbpfgo-test-bpf:
+	$(MAKE) -C $(SELFTEST)/build
+
+libbpfgo-test-bpf-clean:
+	$(MAKE) -C $(SELFTEST)/build clean
+
+# libbpf: shared
+
+libbpfgo-dynamic: outputlibdir
+	CC=$(CLANG) \
+	   CGO_CFLAGS=$(CGO_CFLAGS_DYN) \
+	   CGO_LDFLAGS=$(CGO_LDFLAGS_DYN) \
+	   go build .
+
+libbpfgo-dynamic-install: libbpfgo-dynamic
+	CC=$(CLANG) \
+	   CGO_CFLAGS=$(CGO_CFLAGS_DYN) \
+	   CGO_LDFLAGS=$(CGO_LDFLAGS_DYN) \
+	   go install .
+
+libbpfgo-dynamic-test: libbpfgo-dynamic libbpfgo-test-bpf
+	CC=$(CLANG) \
+	   CGO_CFLAGS=$(CGO_CFLAGS_DYN) \
+	   CGO_LDFLAGS=$(CGO_LDFLAGS_DYN) \
+	   sudo -E go test .
+
+# libbpf: static
+
+libbpfgo-static: $(LIBBPF_OBJ) vmlinuxh
+	CC=$(CLANG) \
+	   CGO_CFLAGS=$(CGO_CFLAGS_STATIC) \
+	   CGO_LDFLAGS=$(CGO_LDFLAGS_STATIC) \
+	   go build .
+
+libbpfgo-static-install: libbpfgo-static
+	CC=$(CLANG) \
+	   CGO_CFLAGS=$(CGO_CFLAGS_STATIC) \
+	   CGO_LDFLAGS=$(CGO_LDFLAGS_STATIC) \
+	   go install .
+
+libbpfgo-static-test: libbpfgo-static libbpfgo-test-bpf
+	CC=$(CLANG) \
+	   CGO_CFLAGS=$(CGO_CFLAGS_STATIC) \
+	   CGO_LDFLAGS=$(CGO_LDFLAGS_STATIC) \
+	   sudo -E go test .
+
+# vmlinux header file
+
+vmlinuxh: outputdir
+	@if [ ! -f $(BTFFILE) ]; then \
```

Comment:
why phony and not a real requirement?

Reply:
Because of the cross-make calls I have in selftest/*/Makefile, for example. The other makefiles (check selftest/common/Makefile) call the main one for their dependencies such as: 

* building the dynamic libbpf
* building static libbpf
* generating vmlinuxh.

This way it works both ways:

`cd ./selftest/perfbuffers ; make`

or 

`make -C selftest/perfbuffers`

and I don't have to keep track of all hard requirements in other Makefiles.

Reply:
it's a good idea to reuse the common targets in the main makefile. really not a big deal but I didn't understand why it needs to be phony though. if the target here would be `$(OUTUPT)/vmlinux.h` and the other makefiles required the same path, wouldn't it work?

Reply:
NM, you are right. There is no need for PHONY here. Fixing it!